server/fishtest/views.py

```diff
@@ -1168,10 +1168,7 @@ def tests_stats(request):
 
 @view_config(route_name="tests_tasks", renderer="tasks.mak")
 def tests_tasks(request):
-    r_id = request.matchdict["id"]
-    run = request.rundb.runs.find_one(
-        {"_id": ObjectId(r_id)}, {"tasks": 1, "bad_tasks": 1, "results": 1, "args": 1}
-    )
+    run = request.rundb.get_run(request.matchdict["id"])
```

Comment:
Yes the aim was to be able to serve the tasks from a secondary instance.

Reply:
Every instance has his separated cache. The problem is: how the data are refreshed/dropped in the cache.

Reply:
Yes but in a secondary instance the cache will not be updated (which happens in the primary instance).

Reply:
So in a secondary instance the first get_run will indeed pull the run from the db and puts it in the cache. But then it will not see any updates anymore, until it is evicted, which takes some time.

Reply:
On PROD now the `pserve` are using 500MB less memory (10% of total RAM). I wonder if the dropped RAM was used for long time cached runs.

Reply:
I implemented this same idea in the open pr for injecting machines but that was because I wanted to expose the two apis on another

We need to test it's performance as such idea of caching db responses still requires locking and deep copying...

Reply:
The primary instance very regularly updates the runs on the db (this is how the main page works, if you refresh it you see the updates, even if they come from the db and not from the slightly more up to date cache).

The time for eviction is substantial but I don't remember how much (I don't have time to look now).

You are right that a finished run does not need see updates, but it can be revived in which case it will see updates again.

Reply:
My personal preference is to keep the rule "get_run = primary instance" and to implement a more primitive time limited cache for secondary instances, if that is deemed necessary (like we do for the main page).

Reply:
Yes this makes sense. I can wait the machine cache performance.. And implement it for tasks.

Reply:
What about renaming `get_run()` is something more clear, eg `get_cached_run()` ?

Reply:
Hmm it does get run from either db or cache so that would be abit not accurate ... I suggest moving it out of the db class. rundb class should only return db stuff..